community/kernel/src/test/java/org/neo4j/kernel/impl/logging/AsyncLogProviderTest.java

```diff
@@ -0,0 +1,198 @@
+        }} );
+        data.add( new Object[] {Level.INFO, new Function<Log,Logger>()
+        {
+            @Override
+            public Logger apply( Log log )
+            {
+                return log.infoLogger();
+            }
+        }} );
+        data.add( new Object[] {Level.WARN, new Function<Log,Logger>()
+        {
+            @Override
+            public Logger apply( Log log )
+            {
+                return log.warnLogger();
+            }
+        }} );
+        data.add( new Object[] {Level.ERROR, new Function<Log,Logger>()
+        {
+            @Override
+            public Logger apply( Log log )
+            {
+                return log.errorLogger();
+            }
+        }} );
+        return data;
+    }
+
+    @Parameter( 0 )
+    public Level level;
+
+    @Parameter( 1 )
+    public Function<Log,Logger> loggerFunction;
+
+    @Test
+    public void shouldProcessLogEventsSeparately() throws Exception
+    {
+        // GIVEN
+        BufferingLogProvider actual = new BufferingLogProvider();
+        AsyncLogProvider asyncLogProvider = new AsyncLogProvider( actual );
+        Log log = asyncLogProvider.getLog( AsyncLogProviderTest.class );
+        Logger logger = log.logger( level );
+
+        logger.log( "First message" );
+        logger.log( "Second %s", "message" );
+        logger.log( "", new RuntimeException( "Third message" ) );
+        BufferingLog actualLog = actual.getLog( AsyncLogProviderTest.class );
+        assertEquals( 0, actualLog.toString().length() );
+
+        // WHEN
+        flush( asyncLogProvider );
+
+        // THEN
+        assertLogContains( actualLog,
+                "First message",
+                "Second message",
+                "Third message",
+                RuntimeException.class.getSimpleName() );
+    }
+
+    @SuppressWarnings( "unchecked" )
+    @Test
+    public void shouldContinueToLogAfterOomDuringLogStatement() throws Exception
+    {
+        // GIVEN
+        ThrowingLogProvider actualLogProvider = new ThrowingLogProvider();
+        Listener<Throwable> errorHandler = mock( Listener.class );
+        final AsyncLogProvider asyncLogProvider = new AsyncLogProvider( actualLogProvider, errorHandler );
+        startLogProcessing( asyncLogProvider );
+
+        // Make sure that the setup in this test is working as it should
+        Log log = asyncLogProvider.getLog( getClass() );
+        log.info( "Testing" );
+        waitForLogMessage( actualLogProvider, "Testing" );
+
+        // WHEN making the actual log provider, i.e. the processing of the log event throw some sort of error
+        actualLogProvider.makeNextLogStatementThrow();
+        log.info( "Should throw" );
+
+        // THEN
+        log.info( "After error" );
+        waitForLogMessage( actualLogProvider, "After error" );
+        assertFalse( actualLogProvider.hasLogged( "Should throw" ) );
+        verify( errorHandler, times( 1 ) ).receive( any( Throwable.class ) );
+    }
```

Comment:
That's probably what we want, but not really possible w/o some refactorings in the logging framework. What we could do is to note the timestamp for when the event was created and compare with when it's logged. If there's a > threshold difference there we can add an additional log statement saying something like `BLOCKED for X ms`.